Check redundancy also for groups that are up

hmusum opened the pull request.

Commit:
Separate code at a higher level based on groupes setup or not

Commit:
Split out method for finding min replication per distributor

Commit:
Check redundancy also for groups that are up

When we allow several groups to go down for maintenance we should check
nodes in the groups that are up if they have the required redundancy. They
might be up but have not yet synced all buckets after coming up. We want
to wait with allowing more nodes to be taken down until that is done.

Commit:
Avoid code duplication

Commit:
Define valid range for groups-allowed-down-ratio in XML schema

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -396,26 +421,55 @@ private Result checkIfNodesAreUpOrRetired(ClusterState clusterState) {
+    }
+
+    private Result checkRedundancySeenFromDistributor(DistributorNodeInfo distributorNodeInfo, Set<Integer> indexesToCheck) {
+        Map<Integer, Integer> replication = new LinkedHashMap<>(minReplication(distributorNodeInfo));
```
Do we need to copy the returned map here?

hmusum replied:
No, can't see why that's needed

vekterli commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -196,86 +209,98 @@ private static Result checkIfStateSetWithDifferentDescription(NodeInfo nodeInfo,
+            for (var n : nodesInGroup) {
+                log.log(FINE, "Check min replication for index " + n.index() + " in group " + group);
+                Set<Integer> indexesToCheck = nodesInGroup.stream().map(ConfiguredNode::index).collect(Collectors.toSet());
+                var r = checkRedundancySeenFromDistributor(clusterInfo.getDistributorNodeInfo(n.index()), indexesToCheck);
```
Are we guaranteed to check _all_ possible distributors that can have knowledge of buckets in the `indexesToCheck` set? Since we prune nodes from groups that are retired/not up. Bucket ownership and management happens _across_ groups and retired-status only applies to content nodes and not distributors, so I'm wondering if we'll risk skipping looking at the corresponding distributors on retired nodes. E.g. if we have 4 groups and 1 has all retired content nodes, 25% of the buckets will still be owned by (non-retired) distributors in that group.

Perhaps we should instead loop over all distributors that are in an available state rather than doing this by group?

hmusum replied:
Yes, you might right about that, will look into it

hmusum replied:
@vekterli Gave it a try, please have a look again

vekterli reviewed (dismissed):
:+1: primarily have a question related to aggregation of replication stats and whether we will always check all relevant distributors

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -396,26 +421,55 @@ private Result checkIfNodesAreUpOrRetired(ClusterState clusterState) {
         return allow();
     }
 
+    private Map<Integer, Integer> minReplication(DistributorNodeInfo distributorNodeInfo) {
```
Yes, that would be nice

Commit:
Fix minor issues after code reviews

hmusum commented on clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java:
```diff
@@ -153,16 +156,22 @@ private Result canSetStateMaintenanceTemporarily(StorageNodeInfo nodeInfo, Clust
-        if (maxNumberOfGroupsAllowedToBeDown == -1) {
-            result = checkIfAnotherNodeInAnotherGroupHasWantedState(nodeInfo);
+        if (isGroupedSetup()) {
+            if (maxNumberOfGroupsAllowedToBeDown == - 1) {
```
Fixed

hmusum commented on clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeCheckerTest.java:
```diff
@@ -215,6 +216,8 @@ void testMaintenanceAllowedFor2Of4Groups() {
 
     @Test
     void testMaintenanceAllowedFor2Of4Groups8Nodes() {
+        // LogSetup.initVespaLogging("foo");
```
Yup, removed

Commit:
Check min replication seen from all distributors that are UP

This means we will also check distributors that are on same node as
a retired storage node

Commit:
Use cluster state to check if disttributors are UP

vekterli reviewed (approved):
:+1:
